### ISSUE TITLE
[BUG][Fuzzer][ice-on-valid-code] `T.shfl_*` on `float16`/`bfloat16` fails to compile instead of shuffling the value

### ISSUE THREAD

**Issue by Hughshine** [tile-ai/tilelang#2569]
### Required prerequisites

- [x] I have read the documentation <https://tilelang.com>.
- [x] I have searched the [Issue Tracker](https://github.com/tile-ai/tilelang/issues) that this hasn't already been reported. (comment there if it has.)

### What version of TileLang are you using?

`0.1.11` (source pinned to tag `v0.1.11`, commit `cd37ed5fc35ae7a60a1277c8eb49028174ac51e6`).

### System information

- Installed from the `0.1.11` wheel; source pinned to tag `v0.1.11` (commit `cd37ed5fc35ae7a60a1277c8eb49028174ac51e6`).
- Python 3.12.3, linux; PyTorch 2.12.1+cu130 (CUDA 13.0).
- GPU: NVIDIA H100 80GB HBM3 (`sm_90`). The defect is in the CUDA codegen emit of `__shfl_*_sync` — a shared, target-independent lowering path (not arch-gated), so the same compile failure is expected on any CUDA target that reaches it; only `sm_90` was run this session.

```python
import sys, tilelang, torch
print(sys.version.split()[0], sys.platform)                              # 3.12.3 linux
print(tilelang.__version__)                                              # 0.1.11
print(torch.__version__, torch.version.cuda)                             # 2.12.1+cu130 13.0
print(torch.cuda.get_device_name(), torch.cuda.get_device_capability())  # NVIDIA H100 80GB HBM3 (9, 0)
```

### Problem description

`T.shfl_xor` / `T.shfl_down` / `T.shfl_up` / `T.shfl_sync` on `float16`/`bfloat16` fail to compile (`nvcc` error) when the shuffle result is assigned to a materialized temporary — because the CUDA codegen emits the raw `__shfl_*_sync` builtin directly ([codegen_cuda.cc#L3986-L4009](https://github.com/tile-ai/tilelang/blob/cd37ed5fc35ae7a60a1277c8eb49028174ac51e6/src/cuda/codegen/codegen_cuda.cc#L3986-L4009)) instead of routing through the `tl::shfl_*_sync` C++ wrappers ([common.h#L955-L973](https://github.com/tile-ai/tilelang/blob/cd37ed5fc35ae7a60a1277c8eb49028174ac51e6/src/tl_templates/cuda/common.h#L955-L973)) whose 16-bit-float specializations exist precisely to make this legal.

For a 16-bit-float operand the raw builtin binds to the `float` overload and **returns `float`**, so `bfloat16_t r = __shfl_xor_sync(...)` (a copy-initialization) hits `cutlass::bfloat16_t`'s missing `float` constructor:

```
tvm_kernels.cu(22): error: no suitable constructor exists to convert from "float" to "cutlass::bfloat16_t"
    bfloat16_t r = __shfl_xor_sync((unsigned int)4294967295, v, 1, 32);
```

The `tl::shfl_*_sync` wrappers ([`half_t`](https://github.com/tile-ai/tilelang/blob/cd37ed5fc35ae7a60a1277c8eb49028174ac51e6/src/tl_templates/cuda/common.h#L977-L1009), [`bfloat16_t`](https://github.com/tile-ai/tilelang/blob/cd37ed5fc35ae7a60a1277c8eb49028174ac51e6/src/tl_templates/cuda/common.h#L1013-L1047)) already do the correct `raw16 → uint32 → __shfl_*_sync → raw16` round-trip and return the 16-bit-float type — but the codegen path never calls them, so those specializations are dead code on the `T.shfl_*` path. The 16-bit float types are the only ones that break: `int32`/`int8`/`float32` were tested and compile with the correct lane permutation (see matrix below), and the other native/32-bit-plus types (`int16`/`int64`/`uint32`/`float64`) are expected to compile for the same reason — the raw builtin has native or safe-promoting overloads for them, and only fp16/bf16 lack a lossless `float`→type conversion.

The failure is code-shape dependent, and differently so for the two 16-bit float types:

<details><summary>Trigger / control matrix (TESTED on H100 sm_90 this session)</summary>

| form | `int32`/`int8`/`fp32` | `float16` | `bfloat16` |
|---|---|---|---|
| materialized temp (`r = T.shfl_xor(v,1); O[tx]=r`) | MATCH | **fails to compile** | **fails to compile** |
| inline store (`O[tx] = T.shfl_xor(A[tx],1)`) | MATCH | MATCH | **fails to compile** |

- **`bfloat16`**: fails in **both** forms — there is no `float`↔`bfloat16_t` conversion at all, so even the inline store fails (`error: no operator "=" matches these operands`).
- **`float16`**: fails only in the temp form. The inline `O[tx] = __shfl_xor_sync(...)` compiles because assigning a `float` result into a `half_t` element is an *implicit* conversion (`half_t` has a `float` constructor), whereas the temp `half_t r = __shfl_xor_sync(...)` is copy-initialization from a `float` that here does not resolve. **The inline-`float16` path that compiles is numerically correct** — verified as an exact lane permutation (lane `L` receives lane `L^1`, 32/32 exact, matching the `float32` control), so this is a **compile-time failure only, not a silent miscompute**; the value round-trips losslessly for exactly-representable inputs.

The matrix above was run with `shfl_xor`. The other three ops (`shfl_down`, `shfl_up`, `shfl_sync`) share the identical raw-emit branch in the same codegen block ([L3986-L4009](https://github.com/tile-ai/tilelang/blob/cd37ed5fc35ae7a60a1277c8eb49028174ac51e6/src/cuda/codegen/codegen_cuda.cc#L3986-L4009) — each emits `__shfl_*_sync` directly with no wrapper routing), so they are expected to fail identically for `fp16`/`bf16` in the temp form for the same reason.

</details>

### Reproducible example code

```python
import tilelang
import tilelang.language as T
import torch


def make_temp(dt):     # materialized-temp form: r = shfl(...); O[tx] = r
    @T.prim_func
    def k(A: T.Tensor((32,), dt), O: T.Tensor((32,), dt)):
        with T.Kernel(1, threads=32):
            tx = T.get_thread_binding()
            v = A[tx]
            r = T.shfl_xor(v, 1, 32)
            O[tx] = r
    return k


def make_inline(dt):   # inline-store form: O[tx] = shfl(A[tx], ...)
    @T.prim_func
    def k(A: T.Tensor((32,), dt), O: T.Tensor((32,), dt)):
        with T.Kernel(1, threads=32):
            tx = T.get_thread_binding()
            O[tx] = T.shfl_xor(A[tx], 1, 32)
    return k


def run(kernel_fn, dt):
    try:
        kk = tilelang.compile(kernel_fn(dt), target="cuda")
    except Exception as e:
        line = next((l.strip() for l in str(e).splitlines() if "error:" in l), type(e).__name__)
        return f"COMPILE ERROR :: {line[:110]}"
    A = (torch.arange(32, dtype=torch.float64).to(getattr(torch, dt)).cuda()
         if "float" in dt else torch.arange(32, dtype=getattr(torch, dt), device="cuda"))
    O = torch.full((32,), -99, dtype=torch.int64).to(getattr(torch, dt)).cuda()
    kk(A, O); torch.cuda.synchronize()
    got = O.to(torch.int64).tolist()
    ref = [L ^ 1 for L in range(32)]         # shfl_xor(v, 1): lane L receives lane L^1
    return "MATCH" if got == ref else f"MISMATCH got={got[:6]} ref={ref[:6]}"


print("temp  int32   :", run(make_temp,   "int32"))     # -> MATCH   (control)
print("temp  float16 :", run(make_temp,   "float16"))   # -> COMPILE ERROR
print("temp  bfloat16:", run(make_temp,   "bfloat16"))  # -> COMPILE ERROR
print("inline float16:", run(make_inline, "float16"))   # -> MATCH   (control)
print("inline bfloat16:", run(make_inline, "bfloat16")) # -> COMPILE ERROR
```

### Traceback

```pytb
<details><summary>nvcc error (bfloat16, temp form) — the emitted line and the diagnostic</summary>


// emitted CUDA (codegen_cuda.cc), materialized temp:
bfloat16_t r = __shfl_xor_sync((unsigned int)4294967295, v, 1, 32);

tvm_kernels.cu(22): error: no suitable constructor exists to convert from "float" to "cutlass::bfloat16_t"
    bfloat16_t r = __shfl_xor_sync((unsigned int)4294967295, v, 1, 32);
1 error detected in the compilation of "tvm_kernels.cu".


For `bfloat16` inline it is instead `error: no operator "=" matches these operands` on
`O[...] = __shfl_xor_sync(...)`. For the working controls the emit is the native path,
e.g. `int r = __shfl_xor_sync((uint)4294967295, v, 1, 32);` (compiles).

</details>
```

### Expected behavior

`T.shfl_*` should compile and return the shuffled 16-bit-float value for `float16`/`bfloat16`, regardless of whether the result is stored inline or through a temporary — the same as it already does for the integer and `float32`/`float64` dtypes. The project already ships the correct machinery for this: the `tl::shfl_*_sync` `half_t`/`bfloat16_t` specializations in `common.h` (added by #1169 specifically to support 16-bit shuffle) do the bit-round-trip and return the right type; the codegen simply does not call them.

### Additional context

**Suggested fix.** Have the `shfl_*` emit call the `tl::shfl_*_sync` wrapper (which already handles the 16-bit-float packing and matches how `reduce.h` uses it) instead of the raw `__shfl_*_sync` builtin — e.g. emit `tl::shfl_xor_sync(mask, value, lane, width)`. Verified in a standalone `nvcc` probe on sm_90 against the shipped cutlass headers: the temp assignment `bfloat16_t r = tl::shfl_xor_sync(mask, v, 1);` compiles cleanly, whereas the raw-builtin form gives the `no suitable constructor` error above. (Codegen is compiled C++, so a shipped fix needs a rebuild; this is a proposed fix probed at the nvcc level, not built end-to-end.)

**Provenance.** The `T.shfl_*` TIR builtins and this raw-`__shfl_*_sync` codegen emit were both introduced together in [#1858](https://github.com/tile-ai/tilelang/pull/1858) ("Add TIR builtins for warp-level vote and block-level predicate sync", merged 2026-04-13). The `tl::shfl_*_sync` 16-bit-float wrappers that would fix it predate it: they were added in [#1169](https://github.com/tile-ai/tilelang/pull/1169) ("Support 16bits shfl_sync", merged 2025-10-31) and wired into `reduce.h`, but #1858's new codegen path did not route through them. The codegen file was later moved (`src/target/` → `src/cuda/codegen/`) by refactors #2121/#2297 — those only relocated the emit, they did not introduce it.

**Relationship to #1162 / #1169 (same root, different site).** This is the *same* `no suitable constructor exists to convert from "float" to "cutlass::bfloat16_t"` failure that [#1162](https://github.com/tile-ai/tilelang/issues/1162) reported for `T.reduce_max` on `bfloat16` — there it fired on the `__shfl_down_sync` inside `reduce.h`. #1169 fixed *that* site (and closed #1162) by adding the `tl::shfl_*_sync` 16-bit-float wrappers and routing `reduce.h` through them. This report is a **fresh occurrence of the same root cause on a new, un-routed path**: #1858's `T.shfl_*` codegen emits the raw builtin directly and never calls those wrappers, so `reduce.h` is fixed while the user-facing `T.shfl_*` builtins are not. So this is not a re-report of #1162 (which is fixed) — it is the identical defect resurfacing where #1169's fix was never applied.

I searched the open and closed tracker (`shfl`, `shuffle`, `bfloat16 shuffle`, `__shfl_xor_sync`, `warp shuffle half`, `no suitable constructor bfloat16`) and found no existing report of this defect on the `T.shfl_*` path. It is distinct from #2478 (`T.atomic_addx4` on fp16/bf16 — a different op and codegen site).
